File: flutter/lib/model/move_feedback.dart
import 'dart:async';

/// A UI operation identity, independent of the provider's changing mail UID.
class MoveRecord {
  MoveRecord(this.id, this.account, this.originalFolder);
  String id;
  final String account, originalFolder;
  bool started = false, cancelled = false, committed = false;
  bool undoRequested = false, blocked = false, restoreCommitted = false;
}

/// The desktop's six-second, destination-scoped move notification contract.
/// Records captured by in-flight work outlive dismissal, but cannot revive it.
class MoveFeedback {
  MoveFeedback(this.changed);
  final void Function() changed;
  Timer? _timer;
  String? _folder, _account;
  bool _restored = false;
  List<MoveRecord> _records = [];
  List<MoveRecord> get records => _records;
  bool get visible => _records.isNotEmpty;
  bool get canUndo => !_restored && _records.any((r) => !r.blocked);
  String? get label {
    if (!visible) return null;
    final count = _records.length, noun = count == 1 ? 'message' : 'messages';
    if (_restored) return 'Restored $count $noun';
    return switch (_folder?.toLowerCase()) {
      'archive' => 'Archived $count $noun',
      'trash' => 'Deleted $count $noun',
      'inbox' => 'Moved $count $noun to Inbox',
      _ => 'Moved $count $noun to $_folder',
    };
  }

  void _schedule() {
    _timer?.cancel();
    _timer = Timer(const Duration(seconds: 6), dismiss);
  }

  MoveRecord add(String id, String account, String original, String folder) {
    final standard = ['archive', 'trash'].contains(folder.toLowerCase());
    final sameFolder = standard
        ? _folder?.toLowerCase() == folder.toLowerCase()
        : _folder == folder;
    if (_restored || !sameFolder || (!standard && _account != account)) {
      _records = [];
    }
    _folder = folder;
    _account = account;
    _restored = false;
    final record = MoveRecord(id, account, original);
    _records = [..._records, record];
    _schedule();
    return record;
  }

  List<MoveRecord> restore(List<MoveRecord> expected) {
    if (!canUndo || !identical(expected, _records)) return [];
    final accepted = _records
        .where((r) => !r.blocked)
        .toList()
        .reversed
        .toList();
    for (final record in accepted) {
      record.undoRequested = true;
      if (!record.started) record.cancelled = true;
    }
    _records = accepted;
    _restored = true;
    _schedule();
    return accepted;
  }

  void failed(MoveRecord record) {
    if (!_records.contains(record)) return;
    _records = _records.where((r) => !identical(r, record)).toList();
    if (_records.isEmpty) _timer?.cancel();
  }

  void retryRestore(MoveRecord record) {
    if (!_restored) _records = [];
    _restored = true;
    if (!_records.contains(record)) _records = [..._records, record];
    _schedule();
  }

  void removeAccount(String account) {
    if (!_records.any((r) => r.account == account)) return;
    _records = _records.where((r) => r.account != account).toList();
    if (_records.isEmpty) _timer?.cancel();
  }

  void dismiss() {
    _timer?.cancel();
    _records = [];
    changed();
  }

  void dispose() => _timer?.cancel();
}

File: web/src/move-feedback.ts
export class MoveRecord {
  started = false;
  cancelled = false;
  committed = false;
  undoRequested = false;
  blocked = false;
  restoreCommitted = false;
  constructor(
    public id: string,
    readonly account: string,
    readonly originalFolder: string,
    /** The other account a transfer moved this message to. */
    readonly destination?: string,
  ) {}
  /** Fields that put the message back where it was. */
  get restoreFields(): { folder: string; accountId?: string } {
    return {
      folder: this.originalFolder,
      ...(this.destination ? { accountId: this.account } : {}),
    };
  }
}

/** Six-second desktop move feedback. Provider completion never recreates it. */
export class MoveFeedback {
  private timer?: ReturnType<typeof setTimeout>;
  private folder?: string;
  private account?: string;
  private restored = false;
  records: MoveRecord[] = [];
  constructor(private changed: () => void) {}
  get visible() {
    return this.records.length > 0;
  }
  get canUndo() {
    return !this.restored && this.records.some((r) => !r.blocked);
  }
  get label() {
    if (!this.visible) return null;
    const count = this.records.length,
      noun = count === 1 ? "message" : "messages";
    if (this.restored) return `Restored ${count} ${noun}`;
    if (this.folder?.toLowerCase() === "archive")
      return `Archived ${count} ${noun}`;
    if (this.folder?.toLowerCase() === "trash")
      return `Deleted ${count} ${noun}`;
    const folder =
      this.folder?.toLowerCase() === "inbox" ? "Inbox" : this.folder;
    return `Moved ${count} ${noun} to ${folder}`;
  }
  private schedule() {
    clearTimeout(this.timer);
    this.timer = setTimeout(() => this.dismiss(), 6000);
  }
  add(
    id: string,
    account: string,
    original: string,
    folder: string,
    destination?: string,
  ) {
    const target = destination ?? account;
    const standard = ["archive", "trash"].includes(folder.toLowerCase());
    const same = standard
      ? this.folder?.toLowerCase() === folder.toLowerCase()
      : this.folder === folder;
    if (this.restored || !same || (!standard && this.account !== target))
      this.records = [];
    this.folder = folder;
    this.account = target;
    this.restored = false;
    const record = new MoveRecord(id, account, original, destination);
    this.records = [...this.records, record];
    this.schedule();
    return record;
  }
  restore(expected: MoveRecord[]) {
    if (!this.canUndo || expected !== this.records) return [];
    const records = this.records.filter((r) => !r.blocked).reverse();
    for (const record of records) {
      record.undoRequested = true;
      if (!record.started) record.cancelled = true;
    }
    this.records = records;
    this.restored = true;
    this.schedule();
    return records;
  }
  failed(record: MoveRecord) {
    if (!this.records.includes(record)) return;
    this.records = this.records.filter((r) => r !== record);
    if (!this.visible) clearTimeout(this.timer);
  }
  retryRestore(record: MoveRecord) {
    if (!this.restored) this.records = [];
    this.restored = true;
    if (!this.records.includes(record))
      this.records = [...this.records, record];
    this.schedule();
  }
  removeAccount(account: string) {
    const involved = (r: MoveRecord) =>
      r.account === account || r.destination === account;
    if (!this.records.some(involved)) return;
    this.records = this.records.filter((r) => !involved(r));
    if (!this.visible) clearTimeout(this.timer);
  }
  dismiss() {
    clearTimeout(this.timer);
    this.records = [];
    this.changed();
  }
  dispose() {
    clearTimeout(this.timer);
  }
}
